### PR TITLE
Use paginated non virtual list in Settings Editor

### PR DESCRIPTION
**Pull request by 9at8**
## Why

#97567 

There are some issues with the virtual list that this navigation model addresses:
- Not all settings are rendered on the page, and as a result, screen reader users cannot use browse mode to navigate settings. (For example, you can press <kbd>h</kbd> or <kbd>Shift + h</kbd> to go to the next or previous heading on the page)
- When users jump from one category to the other, it's not clear that they are in another category. (example tabbing from the last setting in "Commonly Used" to the first setting in "Text editor")
- It's a really long list, and users can get lost in this list

How this control fixes things:
- All settings are rendered on the page. (currently, the page size is 20)
- Only the settings related to a particular category are rendered, so users can't jump to another category.
- It's paginated, and there are only 20 settings per page for now.

This control could only be enabled when in screen reader mode, or under a setting, **but** maintaining two kinds of control is harder than one. This is why I want to merge this in insiders for all users and gather feedback for this change. (both a11y related and non-a11y related feedback)

## Gifs

recording (4)
recording (5)
recording (6)

### CODE REVIEW

**Comment by roblourens** [comment]
This is weird, sometimes the scrollbar transition stops working for me

- Open the settings editor
- Click away to some other window
- Click back into vscode
- Now the scrollbar doesn't fade out, it disappears instantly when I stop hovering
- Pick a different category
- Now the transition takes effect again

Any idea why?

**Comment by roblourens** [comment]
image

The wrapped text looks too tight, I think we should tweak the lineHeight

**Comment by roblourens** [comment]
image

I think only add the `...` if you are omitting more than one number

btw these are minor issues so far, we can fix them after merging

**Comment by roblourens** [comment]
Also what happened to using a border on the selected page?

**Comment by roblourens** [comment]
I keep seeing this weird glitch, but I see it in Insiders too, do you happen to know what it is? 
image

File: src/vs/workbench/contrib/preferences/browser/settingsList.ts [hunk 1]
```diff
@@ -0,0 +1,437 @@
...
+	}
+
+	private renderPage(shouldScroll: boolean): void {
+		DOM.clearNode(this.container);
+		this.pageDisposables.clear();
+
+		// Transfer all used items to the free pool
+		for (const [templateId, usedItems] of this.usedPool.entries()) {
+			const freeItems = this.freePool.get(templateId) ?? [];
+			usedItems.forEach(item => item.template.elementDisposables.clear());
```
Comment (roblourens) [inline src/vs/workbench/contrib/preferences/browser/settingsList.ts:187]: I think you should call `disposeElement` on the renderer instead. This level shouldn't know anything about what's inside a template.

File: src/vs/workbench/contrib/preferences/browser/settingsList.ts [hunk 2]
```diff
@@ -0,0 +1,437 @@
...
+}
+
+interface ISettingsListCacheItem {
+	container: HTMLElement;
+	template: ISettingItemTemplate;
+}
+
+export class SettingsList extends Disposable {
+	private searchFilter: (element: SettingsTreeElement) => boolean;
+	private getTemplateId = new SettingsTreeDelegate().getTemplateId;
```
Comment (roblourens) [inline src/vs/workbench/contrib/preferences/browser/settingsList.ts:88]: Not a safe way to access a method on an instance. You would have to call `.bind()` with the instance. But you should just hang on to the SettingsTreeDelegate itself

File: src/vs/workbench/contrib/preferences/browser/settingsList.ts [hunk 3]
```diff
@@ -0,0 +1,437 @@
...
+
+	setSettings(settings: SettingsTreeSettingElement[], scrollToPage?: number): void {
+		this.settings = settings;
+		this.setPage(scrollToPage ?? this.page);
+	}
+
+	nextPage(): void {
+		const nextStartIdx = this.page * this.PAGE_SIZE;
+
+		if (this.settings.length > nextStartIdx) {
```
[the remark this answers is not in the document]
Comment (9at8) [reply src/vs/workbench/contrib/preferences/browser/settingsList.ts:56]: You're right, not needed.

File: src/vs/workbench/contrib/preferences/browser/settingsList.ts [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+			if (isDefined(result)) {
+				return result;
+			}
+		}
+	}
+
+	return;
+}
+
+function themingParticipant(theme: IColorTheme, collector: ICssStyleCollector) {
```
[the remark this answers is not in the document]
Comment (9at8) [reply src/vs/workbench/contrib/preferences/browser/settingsList.ts:367]: Yep, that's correct.

**Comment by 9at8** [comment]
> I keep seeing this weird glitch, but I see it in Insiders too, do you happen to know what it is?

I have never seen it :thinking:

**Review by roblourens — APPROVED** [review]

**Pull request merged**